### PR TITLE
Per-program id timings

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Hard to tell which program is taking the longest in the network.

#### Summary of Changes

Add per-program timings configurable with a validator flag.

Fixes #

### CODE REVIEW

File: core/src/progress_map.rs [hunk 1]
```diff
@@ -104,6 +104,27 @@ impl ReplaySlotStats {
...
+
+        let mut per_pubkey_timings: Vec<_> = self
+            .execute_timings
+            .details
+            .per_program_timings
+            .iter()
+            .collect();
+        per_pubkey_timings.sort_by(|a, b| b.1.cmp(&a.1));
+        let total: u64 = per_pubkey_timings.iter().map(|a| a.1).sum();
+        for (pubkey, time) in per_pubkey_timings.iter().take(5) {
```
Comment (taozhu-chicago) [inline core/src/progress_map.rs:116]: Perhaps to remove `take(5)` when utilize this change to feed cost model?

File: core/src/validator.rs [hunk 2]
```diff
@@ -193,6 +194,7 @@ impl Default for ValidatorConfig {
             tpu_coalesce_ms: DEFAULT_TPU_COALESCE_MS,
             validator_exit: Arc::new(RwLock::new(ValidatorExit::default())),
             no_wait_for_vote_to_start_leader: true,
+            collect_per_program_timings: false,
```
Comment (taozhu-chicago) [inline core/src/validator.rs:197]: default it to `true`?

File: core/src/progress_map.rs (hunk 1 above)
Comment (sakridge) [reply core/src/progress_map.rs:116]: Well this is just to report metrics for the top 5 costly programs. So I think we would leave this in. There would just be another code-path to update the cost model with these values.

Comment (taozhu-chicago) [reply core/src/progress_map.rs:116]: right, makes sense, this function just reports stats. Will add code at replay_stage to read execute_timings for cost model

File: runtime/src/message_processor.rs [hunk 3]
```diff
@@ -1222,28 +1228,36 @@ impl MessageProcessor {
...
+                    timings,
+                    demote_sysvar_write_locks,
+                    account_db.clone(),
+                    ancestors,
+                )
+                .map_err(|err| TransactionError::InstructionError(instruction_index as u8, err));
+            time.stop();
+            if timings.collect_per_program_timings {
+                let program_id = instruction.program_id(&message.account_keys);
+                *timings.per_program_timings.entry(*program_id).or_insert(0) += time.as_us();
```
Comment (taozhu-chicago) [inline runtime/src/message_processor.rs:1258]: In the case of same program being executed multiple times in a transaction, do we want to see the accumulated execution time in stats report, or perhaps averaged time? 
Cost model interested in later, if reporting accumulated time is desired, it is always possible to take average at the time of updating cost model.

Comment (sakridge) [reply runtime/src/message_processor.rs:1258]: yes. I think we'll have to have a count stat here as well. I'll add that.

File: core/src/validator.rs (hunk 2 above)
Comment (sakridge) [reply core/src/validator.rs:197]: I'll just remove it, if we need it for the cost model then we need to do it all the time anyway.

**Review by taozhu-chicago — APPROVED** [review]
looks good

**Pull request merged**